app-core/src/main/java/net/spals/appbuilder/app/core/generic/GenericWorkerApp.java

```diff
@@ -105,6 +124,7 @@ public Builder setServiceScan(final ServiceScan serviceScan) {
             addBootstrapModule(new AutoBindModulesBootstrapModule(serviceScan));
             // 3. Use the serviceScan to find auto bound services
             servicesModuleBuilder.setServiceScan(serviceScan);
+            servicesModuleIndex.set(orderedModules.size());
```

Comment:
The ultimate goal is to create an `Injector` instance. To do that, we're creating a series of `Module`s (many `Module`s together create an `Injector`). `Module`s come in two flavors: custom and standard. Incidentally, there are `BootstrapModule`s too, but we'll ignore that now for the purposes of the explanation.

Custom `Module`s are completely defined by the application author and are added via the `addModule(Module)` method.

Standard `Module`s are created internally by this builder class based on state provided to it from the application author. There's actually only one standard `Module`: `AutoBindServicesModule`. What this does is scan your code and automatically binds annotated classes into the Guice `Injector`. `AutoBindServicesModule` requires two pieces of state: a `ServiceScan` and the boolean flag `errorOnServiceLeaks` (which is defaulted to true, but you can turn it off via `disableErrorOnServiceLeaks`). `AutoBindServicesModule` itself actually uses FreeBuilder to allow these two pieces of state to arrive in any order. In other words, you can disable the error leaks first and then set the `ServiceScan` or the other way around. It doesn't matter. 

It used to be that, when we gathered up all the `Module`s, we would put the `AutoBindServicesModule` last in the list (after any custom `Module`s provided by the app author). This was perfectly fine because the order of `Module`s didn't matter. The trouble started when I introduced this binding overrides feature (which, incidentally, should really only be enabled for tests). Binding overrides is more permissive in that it allows multiple things to be bound to a singleton interface (normally, Guice would throw an error for this). It gets away with that by picking the __**last**__ class that was bound to the interface. So this means that the order in which `Module`s are added is suddenly very important.

The tricky part is that the `AutoBindServicesModule` isn't available until the app author calls the `.build()` method. But we want it placed in the `Module` list in the order in which it arrived (as determined by the call to `setServiceScan(ServiceScan` which really is the meat of the `AutoBindServicesModule` state). So when you call `setServiceScan(ServiceScan)`, I remember the index that the `AutoBindServicesModule` should be in the ordered `Module` list. You can think of that as a placeholder. Then when the author calls `build()`, I examine the placeholder and insert the now available `AutoBindServicesModule` into the correct position in the ordered `Module`s list. And that is how I maintain the order of `Module`s to support the binding overrides feature.